### PR TITLE
gh-111968: Refactor _PyXXX_Fini to integrate with _PyObject_ClearFreeLists

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-111968

### CODE REVIEW

**Comment by corona10** [comment]
Address by @ericsnowcurrently 's and @colesbury 's review from https://github.com/python/cpython/issues/111968#issuecomment-1922247135

File: Include/internal/pycore_gc.h [hunk 1]
```diff
@@ -262,11 +262,11 @@ extern PyObject *_PyGC_GetReferrers(PyInterpreterState *interp, PyObject *objs);
 
 // Functions to clear types free lists
 extern void _PyGC_ClearAllFreeLists(PyInterpreterState *interp);
-extern void _Py_ClearFreeLists(_PyFreeListState *state, int is_finalization);
+extern void _PyObject_ClearFreeLists(_PyFreeListState *state, int is_finalization);
 extern void _PyTuple_ClearFreeList(_PyFreeListState *state, int is_finalization);
 extern void _PyFloat_ClearFreeList(_PyFreeListState *state, int is_finalization);
 extern void _PyList_ClearFreeList(_PyFreeListState *state, int is_finalization);
-extern void _PySlice_ClearCache(_PyFreeListState *state);
+extern void _PySlice_ClearCache(_PyFreeListState *state, int is_finalization);
```
Comment (colesbury) [inline Include/internal/pycore_gc.h]: I would name this like the other functions `_PySlice_ClearFreeList`

**Review by colesbury — COMMENTED** [review]
This looks good to me, although I'd probably name `PySlice_ClearCache` like the other freelist clearing functions for consistency.

File: Python/pystate.c [hunk 2]
```diff
@@ -1459,7 +1459,7 @@ clear_datastack(PyThreadState *tstate)
 }
 
 void
-_Py_ClearFreeLists(_PyFreeListState *state, int is_finalization)
+_PyObject_ClearFreeLists(_PyFreeListState *state, int is_finalization)
```
Comment (ericsnowcurrently) [inline Python/pystate.c]: FWIW, it seems odd to me that `_PyObject_ClearFreeLists()` is implemented in pystate.c.  I'd expect it in Objects/object.c.  It isn't critical, but consider moving it there.

File: Python/pystate.c [hunk 3]
```diff
@@ -1557,8 +1559,7 @@ PyThreadState_Clear(PyThreadState *tstate)
 #ifdef Py_GIL_DISABLED
     // Each thread should clear own freelists in free-threading builds.
     _PyFreeListState *freelist_state = &((_PyThreadStateImpl*)tstate)->freelist_state;
```
Comment (ericsnowcurrently) [inline Python/pystate.c]:
```suggestion
    _PyFreeListState *freelist_state = _PyFreeListState_GET();
```

File: Include/internal/pycore_gc.h [hunk 4]
```diff
@@ -262,11 +262,11 @@ extern PyObject *_PyGC_GetReferrers(PyInterpreterState *interp, PyObject *objs);
...
 extern void _PyDict_ClearFreeList(_PyFreeListState *state, int is_finalization);
 extern void _PyAsyncGen_ClearFreeLists(_PyFreeListState *state, int is_finalization);
 extern void _PyContext_ClearFreeList(_PyFreeListState *state, int is_finalization);
```
Comment (ericsnowcurrently) [inline Include/internal/pycore_gc.h]: FWIW, I personally would put these function declarations in the respective type-specific header files, e.g. `_PyTuple_ClearFreeList()` in pycore_tuple.h.  That's the first place I'd think to look for a type-specific declaration like this.  The second place I'd look is pycore_freelist.h.  I would not think to look for it in pycore_gc.h.

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
Thanks for doing this.  It mostly looks okay to me.  There are just a few small things I'd like you to consider.

File: Include/internal/pycore_freelist.h [hunk 5]
```diff
@@ -125,6 +125,16 @@ typedef struct _Py_freelist_state {
     struct _Py_object_stack_state object_stacks;
 } _PyFreeListState;
 
+extern void _PyObject_ClearFreeLists(_PyFreeListState *state, int is_finalization);
```
Comment (corona10) [inline Include/internal/pycore_freelist.h:128]: Since this API is not a domain-specific API, I located all ClearFreeList API into here :)

**Comment by corona10** [comment]
I have made the requested changes; please review again

**Comment by corona10** [comment]
As I commented before, we need some header cleanup, but I will do it as a separate PR since it affected a lot of codes, including `bytecodes.c`

**Comment by corona10** [comment]
@ericsnowcurrently gentle ping

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

**Pull request merged**